react/features/whiteboard/middleware.ts

```diff
@@ -0,0 +1,166 @@
+    }
+};
+
+/**
+ * Handles the whiteboard state.
+ *
+ * @param {Store} store - The redux store.
+ * @param {Object} attributes - The command attributes.
+ * @returns {void}
+ */
+const handleWhiteboardStatus = (store: IStore, { roomId, roomKey }: { roomId: string; roomKey: string; }) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const isOpen = isWhiteboardOpen(state);
+
+    if (isOpen) {
+        return;
+    }
+
+    focusWhiteboard(store);
+    dispatch(setupWhiteboard({
+        collabDetails: {
+            roomId,
+            roomKey
+        }
+    }));
+};
+
+/**
+ * Sends the WHITEBOARD command.
+ *
+ * @returns {void}
+ */
+const sendWhiteboardCommand = ({ conference, roomId, roomKey }: {
+    conference: any;
+    roomId?: string;
+    roomKey?: string;
+}): void => {
+    conference.sendCommand(WHITEBOARD, {
+        attributes: {
+            roomId,
+            roomKey
+        }
+    });
+};
+
+/**
+ * Middleware which intercepts whiteboard actions to handle changes to the related state.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register((store: IStore) => (next: Function) => async (action: any) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const localParticipant = getLocalParticipant(state);
+
+    switch (action.type) {
+    case OPEN_WHITEBOARD: {
+        if (!isParticipantModerator(localParticipant) && !isEveryoneModerator(state)) {
+            return;
+        }
+
+        const conference = getCurrentConference(state);
+        const collabDetails = await generateCollaborationLinkData();
+
+        dispatch(setTileView(false));
+        sendWhiteboardCommand({
+            ...collabDetails ? {
+                roomId: getCurrentRoomId(state),
+                roomKey: collabDetails?.roomKey
+            } : {},
+            conference
+        });
+
+        break;
+    }
+    case CONFERENCE_JOIN_IN_PROGRESS: {
+        const { conference } = action;
+
+        // @ts-ignore
+        conference.addCommandListener(WHITEBOARD, ({ attributes }) => {
+            handleWhiteboardStatus(store, attributes);
+        });
+        break;
+    }
+
+    case SHOW_WHITEBOARD: {
+        focusWhiteboard(store);
+
+        break;
+    }
+    case HIDE_WHITEBOARD: {
+        const conference = getCurrentConference(state);
+
+        dispatch(participantLeft(WHITEBOARD_ID, conference));
+        break;
+    }
+    }
+
+    return next(action);
+});
+
+/**
+ * Set up state change listener to perform maintenance tasks when the conference
+ * is left or failed, e.g. Disable the whiteboard if it's left open.
+ */
+StateListenerRegistry.register(
+
+    // @ts-ignore
+    state => getCurrentConference(state),
+
+    // @ts-ignore
+    (conference, { dispatch }, previousConference): void => {
+        if (conference !== previousConference) {
+
+            dispatch(resetWhiteboard());
```

Comment:
If the whiteboard was being shown, shouldn't it be hidden and destroyed?

Reply:
Reset does just that, cleans up its reducer and unmounts the ExcalidrawApp. Do you have something else in mind?

Reply:
Does that trigger a participant left too?

Reply:
Now it does :grimacing: